Branch-SDK/src/main/java/io/branch/referral/BillingGooglePlay.kt

```diff
@@ -0,0 +1,243 @@
+                        revenue,
+                        "Subscription"
+                    )
+                }
+            } else {
+                Log.e(
+                    "BranchEvent",
+                    "Failed to query subscriptions. Error code: " + billingResult.responseCode
+                )
+            }
+        }
+
+        billingClient?.queryProductDetailsAsync(
+            queryProductDetailsParams
+        ) { billingResult: BillingResult, productDetailsList: List<ProductDetails?> ->
+            if (billingResult.responseCode == BillingClient.BillingResponseCode.OK) {
+
+                var contentItemBUOs: MutableList<BranchUniversalObject> =
+                    ArrayList()
+                var currency: CurrencyType = CurrencyType.USD
+                var revenue = 0.00
+                var quantity: Int = purchase.quantity
+
+                for (product: ProductDetails? in productDetailsList) {
+                    val buo: BranchUniversalObject =
+                        createBUOWithInAppProductDetails(product, quantity)
+                    contentItemBUOs.add(buo)
+                    revenue += (buo.contentMetadata.price * quantity)
+                    currency = buo.contentMetadata.currencyType
+                }
+
+                if (contentItemBUOs.isNotEmpty()) {
+                    createAndLongEventForPurchase(
+                        context,
+                        purchase,
+                        contentItemBUOs,
+                        currency,
+                        revenue,
+                        "In-App Purchase"
+                    )
+                }
+            } else {
+                Log.e(
+                    "BranchEvent",
+                    "Failed to query products. Error code: " + billingResult.responseCode
+                )
+            }
+        }
+    }
+
+    private fun createBUOWithSubsProductDetails(product: ProductDetails?): BranchUniversalObject {
+        if (product != null) {
+
+            val pricingPhaseList =
+                product.subscriptionOfferDetails?.get(0)?.pricingPhases?.pricingPhaseList?.get(0)
+
+            val currency = pricingPhaseList?.let {
+                CurrencyType.valueOf(
+                    it.priceCurrencyCode
+                )
+            }
+
+            val price = ((pricingPhaseList?.priceAmountMicros ?: 0) / 1000000).toDouble()
+
+            return BranchUniversalObject()
+                .setCanonicalIdentifier(product.productId)
+                .setTitle(product.title)
+                .setContentMetadata(
+                    ContentMetadata()
+                        .addCustomMetadata("product_type", product.productType)
+                        .setPrice(price, currency)
+                        .setProductName(product.name)
+                        .setQuantity(1.0)
+                        .setContentSchema(BranchContentSchema.COMMERCE_PRODUCT)
+
+                )
+        } else {
```

Comment:
When would it be the case that the `product` is `null`? 
Would we still want to log an event with an empty BUO?